src/modules/azure-azcopy-sync.ts

```diff
@@ -0,0 +1,134 @@
+import { Echo } from "./echo";
+import shell from "shelljs";
+import { CommandStringBuilder } from "../utilities/command-string-builder";
+import child_process from "child_process";
+
+// -----------------------------------------------------------------------------------------
+// #region Constants
+// -----------------------------------------------------------------------------------------
+
+const AZCOPY_COMMAND = "azcopy sync";
+
+// #endregion Constants
+
+// -----------------------------------------------------------------------------------------
+// #region Interfaces
+// -----------------------------------------------------------------------------------------
+
+interface BlobStorageUrlParts {
+    account: string;
+    container: string;
+    path?: string;
+    sasToken?: string;
+}
+
+interface SyncLocalFileOptions {
+    localFilePath: string;
+    destination: BlobStorageUrlParts;
+}
+
+interface SyncContainerOptions {
+    deleteDestination: boolean;
+    destination: BlobStorageUrlParts;
+    recursive: boolean;
+    source: BlobStorageUrlParts;
+}
+
+// #endregion Interfaces
+
+// -----------------------------------------------------------------------------------------
+// #region Functions
+// -----------------------------------------------------------------------------------------
+
+const _getCommandStringBuilder = (
+    source: string,
+    destination: string,
+    flags: string[]
+): CommandStringBuilder => {
+    return new CommandStringBuilder(
+        "azcopy",
+        "sync",
+        source,
+        destination,
+        ...flags
+    );
+};
+
+const _getBlobStorageUrl = (urlParts: BlobStorageUrlParts) => {
+    const path = urlParts.path ?? "";
+    const sasToken = urlParts.sasToken ?? "";
+    return `"https://${urlParts.account}.blob.core.windows.net/${urlParts.container}/${path}${sasToken}"`;
+};
+
+const _getDeleteDestination = (deleteDestination: boolean) => {
+    return `--delete-destination=${deleteDestination}`;
+};
+
+const _getRecursiveFlag = (recursive: boolean) => {
+    return `--recursive=${recursive}`;
+};
+
+const _sync = (
+    source: string,
+    destination: string,
+    action: string,
+    flags: string[]
+) => {
+    const { cmd, args } = _getCommandStringBuilder(source, destination, flags);
+
+    Echo.message(`Starting the action of ${action}`);
+    const { status } = child_process.spawnSync(cmd, args, {
+        stdio: "inherit",
+        shell: true,
+    });
+    if (status != null && status !== 0) {
+        Echo.error(`Failed when performing the action of ${action}`);
+        shell.exit(status);
+    }
+    Echo.success(`Succeeded when performing the action of ${action}`);
+};
+
+const AzureAzcopySync = {
```

Comment:
Do we think `Azure` is necessary context here? For now, we can probably assume the only `Azcopy` utility we're working with is related to Azure. If we begin building out a number of azure-related modules, maybe we could then put them in an `azure` subfolder. (Might not be a bad idea for the existing `dotnet` modules, since there are quite a few now)

Reply:
I feel that if we make this change that we should also be changing the file name to `azcopy-sync.ts`.  If we're going to make the one changes, should make the other change as well.  It's a little confusing because we have an `azure.ts` which, by itself, it responsible for logging in and out of azure using the Azure CLI (using `az`).  AzCopy has it's own CLI (using `azcopy`).  Maybe it makes sense to also have an `azcopy.ts` whic is responsible for it's log in / log out operations.  (see https://docs.microsoft.com/en-us/azure/storage/common/storage-ref-azcopy-login?toc=/azure/storage/blobs/toc.json).  I'll make the change you suggest but won't add anything else.